### PR TITLE
Timex function for hasFullDate and hasValidHour builtin-functions

### PR DESCRIPTION
**Pull request by cosmicshuai**
closes: #2869 
Now we have 7 timex related functions:
HasFullDate, HasValidTime, HasValidDuration, HasValidDate, HasValidTimeRange, HasValidDateRange, IsPresent

### CODE REVIEW

File: libraries/AdaptiveExpressions/ExpressionFunctions.cs [hunk 1]
```diff
@@ -3701,6 +3701,76 @@ private static IEnumerable<object> Flatten(IEnumerable<object> list, int dept)
...
+                            {
+                                parsed = jTimex.ToObject<TimexProperty>();
+                            }
+                            else if (args[0] is string ts)
+                            {
+                                parsed = new TimexProperty(ts);
+                            } 
+                            else
+                            {
+                                error = $"{expr} should be either a TimexProperty or a string";
```
Comment (chrimc62) [inline libraries/AdaptiveExpressions/ExpressionFunctions.cs:3729]: >{expr} should be either a TimexProperty or a string [](start = 42, length = 51)

Not quite the right error string.  $"{expr} requires a TimexProperty or string as an argument."

File: libraries/AdaptiveExpressions/ExpressionFunctions.cs [hunk 2]
```diff
@@ -3701,6 +3701,76 @@ private static IEnumerable<object> Flatten(IEnumerable<object> list, int dept)
...
+                            }
+                        }
+
+                        value = parsed != null && parsed.Year != null && parsed.Month != null && parsed.DayOfMonth != null;
+
+                        return (value, error);
+                    },
+                    ReturnType.Boolean,
+                    expr => ValidateArityAndAnyType(expr, 1, 1, ReturnType.Object)),
```
Comment (chrimc62) [inline libraries/AdaptiveExpressions/ExpressionFunctions.cs:3738]: >ValidateArityAndAnyType(expr, 1, 1, ReturnType.Object) [](start = 28, length = 54)

You can validate to object | string, right?  Numbers or arrays are not allowed.

File: libraries/AdaptiveExpressions/ExpressionFunctions.cs [hunk 3]
```diff
@@ -3701,6 +3701,76 @@ private static IEnumerable<object> Flatten(IEnumerable<object> list, int dept)
...
+                            {
+                                parsed = jTimex.ToObject<TimexProperty>();
+                            }
+                            else if (args[0] is string ts)
+                            {
+                                parsed = new TimexProperty(ts);
+                            }
+                            else
+                            {
+                                error = $"{expr} should be either a TimexProperty or a string";
```
Comment (chrimc62) [inline libraries/AdaptiveExpressions/ExpressionFunctions.cs:3764]: >$"{expr} should be either a TimexProperty or a string"; [](start = 40, length = 55)

same comment

File: libraries/AdaptiveExpressions/ExpressionFunctions.cs [hunk 4]
```diff
@@ -3701,6 +3701,76 @@ private static IEnumerable<object> Flatten(IEnumerable<object> list, int dept)
...
+                            }
+                        }
+
+                        value = parsed != null && parsed.Hour != null;
+
+                        return (value, error);
+                    },
+                    ReturnType.Boolean,
+                    expr => ValidateArityAndAnyType(expr, 1, 1, ReturnType.Object)),
```
Comment (chrimc62) [inline libraries/AdaptiveExpressions/ExpressionFunctions.cs:3773]: >alidateArityAndAnyType(expr, 1, 1, ReturnType.Object)), [](start = 29, length = 55)

same comment

File: tests/AdaptiveExpressions.Tests/BadExpressionTests.cs [hunk 5]
```diff
@@ -172,6 +172,10 @@ public class BadExpressionTests
             #endregion
             
             #region Date and time function test
+            Test("hasFullDate(12345)"), // should hava string or TimexProperty parameter
```
Comment (chrimc62) [inline tests/AdaptiveExpressions.Tests/BadExpressionTests.cs:175]: >hava [](start = 50, length = 4)

have a

File: tests/AdaptiveExpressions.Tests/BadExpressionTests.cs [hunk 6]
```diff
@@ -172,6 +172,10 @@ public class BadExpressionTests
...
+            Test("hasFullDate('world', 123445)"), // should only one parameter
+            Test("hasValidHour(123445)"), // should hava string or TimexProperty parameter
```
Comment (chrimc62) [inline tests/AdaptiveExpressions.Tests/BadExpressionTests.cs:177]: >hava [](start = 52, length = 4)

have a

**Comment by chrimc62** [comment]
Why not hasValidX for other things like month/day/year/minute/duration?  If I said "when do you want to leave" and got a response of "the 30th", I might want to check for not having a month for example.  

---
In reply to: [399361646](https://github.com/microsoft/botbuilder-dotnet/pull/3780#issuecomment-399361646) [](ancestors = 399361646)

File: libraries/AdaptiveExpressions/ExpressionFunctions.cs (hunk 2 above)
Comment (cosmicshuai) [reply libraries/AdaptiveExpressions/ExpressionFunctions.cs:3738]: Yes, these two functions can have a string or a TimexProperty as an argument. 
Here ReturnType.Object represents a base type which includes string, number and boolean, etc.
The actual type checking is in the body of functions.

File: libraries/AdaptiveExpressions/ExpressionFunctions.cs (hunk 1 above)
Comment (cosmicshuai) [reply libraries/AdaptiveExpressions/ExpressionFunctions.cs]: That's more accurate. Thanks

File: tests/AdaptiveExpressions.Tests/BadExpressionTests.cs (hunk 5 above)
Comment (cosmicshuai) [reply tests/AdaptiveExpressions.Tests/BadExpressionTests.cs]: fixed

File: tests/AdaptiveExpressions.Tests/BadExpressionTests.cs (hunk 6 above)
Comment (cosmicshuai) [reply tests/AdaptiveExpressions.Tests/BadExpressionTests.cs]: fixed

**Comment by cosmicshuai** [comment]
@vishwacsena 
Based on Chris' comment above, maybe we can add a 'hasValidTimexProperty' function, which is simillar to addToTime. The signature of this function is hasValidTimexProperty(timex, property). 
timex is either a TimexProperty or a string,
property is a string, like year, month, day, hour, duration, etc.

**Comment by vishwacsena** [comment]
@cosmicshuai yes, we can always take that as an additive change. Are you suggesting we do it instead of this one? If so, I'd only be concerned about scoping it (which is what the trimmed proposal here was supposed to be). My vote would be for us to add these now and add the generic function to the backlog. Ok?

**Comment by cosmicshuai** [comment]
@vishwacsena 
Sure. I will add the generic function in another PR.

**Comment by chrimc62** [comment]
I'm not sure I buy this.  While it lessens the risk for now, it comes at the cost of adding more functions which overlap--doesn't seem like a good tradeoff to me.

---
In reply to: [618771234](https://github.com/microsoft/botbuilder-dotnet/pull/3780#issuecomment-618771234) [](ancestors = 618771234)

**Comment by cosmicshuai** [comment]
@chrimc62 @vishwacsena 
As I found some references about timex inference functions, 
https://github.com/microsoft/Recognizers-Text/blob/master/.NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexInference.cs
I rewrote timex prebuilt functions. Now we have 7 related functions,
HasFullDate, HasValidTime, HasValidDuration, HasValidDate, HasValidTimeRange, HasValidDateRange, IsPresent
The functionalities are identical to those in TimexInference. 
Do you have any idea or suggestions, like should we keep the same naming as the TimexInference library?

File: libraries/AdaptiveExpressions/ExpressionFunctions.cs (hunk 2 above)
Comment (boydc2014) [reply libraries/AdaptiveExpressions/ExpressionFunctions.cs:3750]: Our ReturnType.Object today actually works like a `any` not a `object`, so object | number won't be able to reject number or array, so we probably can refine the type checking types later.

**Comment by chrimc62** [comment]
I like keeping the names the same.

---
In reply to: [620470357](https://github.com/microsoft/botbuilder-dotnet/pull/3780#issuecomment-620470357) [](ancestors = 620470357)

File: tests/AdaptiveExpressions.Tests/BadExpressionTests.cs (hunk 5 above)
Comment (chrimc62) [reply tests/AdaptiveExpressions.Tests/BadExpressionTests.cs]: Still have lots of "hava" in this file that should be "have"

---
In reply to: [414243017](https://github.com/microsoft/botbuilder-dotnet/pull/3780#discussion_r414243017) [](ancestors = 414243017)

**Pull request merged**